polygeist/test/polygeist-opt/invalid_canonicalization.mlir

```diff
@@ -46,3 +46,19 @@ func.func @SimplifySubIndexUsers(%arg0: memref<?x!llvm.struct<(i32)>>) -> memref
 }
 
 // -----
+
+// CHECK:  func.func @Memref2PointerIndex([[A0:%.*]]: memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32> {
+// CHECK-NEXT:    [[C1:%.*]] = arith.constant 1 : index
+// CHECK-NEXT:    [[T0:%.*]] = "polygeist.subindex"([[A0]], [[C1]]) : (memref<?x!llvm.struct<(i32, i32)>>, index) -> memref<?xi32>
+// CHECK-NEXT:    [[T1:%.*]] = "polygeist.memref2pointer"([[T0]]) : (memref<?xi32>) -> !llvm.ptr<i32>
+// CHECK-NEXT:    return [[T1]] : !llvm.ptr<i32>
+// CHECK-NEXT:  }
```

Comment:
I'd argue that we'd ideally want something like:
```mlir
// CHECK:  func.func @Memref2PointerIndex([[A0:%.*]]: memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32> {
// CHECK-NEXT:    [[T0:%.*]] = "polygeist.memref2pointer"([[A0]]) : (memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<?x!llvm.struct<(i32, i32)>>
// CHECK-NEXT:    [[T1:%.*]] = llvm.getelementptr [[T0]][1, 0] : (!llvm.ptr<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32>
// CHECK-NEXT:    return [[T1]] : !llvm.ptr<i32>
// CHECK-NEXT:  }
```
If the types differed, we'd need to do something like:
```mlir
// CHECK:  func.func @Memref2PointerIndex([[A0:%.*]]: memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i8> {
// CHECK-NEXT:    [[T0:%.*]] = "polygeist.memref2pointer"([[A0]]) : (memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<?x!llvm.struct<(i32, i32)>>
// CHECK-NEXT:    [[T1:%.*]] = llvm.getelementptr [[T0]][1] : (!llvm.ptr<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<?x!llvm.struct<(i32, i32)>>
// CHECK-NEXT:    [[T2:%.*]] = llvm.bitcast([[T1]]) : !llvm.ptr<?x!llvm.struct<(i32, i32)>> to !llvm.ptr<i8>
// CHECK-NEXT:    return [[T2]] : !llvm.ptr<i8>
// CHECK-NEXT:  }
```
Can we address this in this PR?

Reply:
Given that `polygeist.subindex` is lowered to `llvm.getelementptr` in `ConvertPolygeistToLLVM` pass, why is it better to duplicate the logic and do this canonicalization here?

Assuming we want to minimize `polygeist.subindex`:
Should the first case be:
```
// CHECK:  func.func @Memref2PointerIndex([[A0:%.*]]: memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32> {
// CHECK-NEXT:    [[T0:%.*]] = "polygeist.memref2pointer"([[A0]]) : (memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<!llvm.struct<(i32, i32)>>
// CHECK-NEXT:    [[T1:%.*]] = llvm.getelementptr [[T0]][0, 1] : (!llvm.ptr<!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32>
// CHECK-NEXT:    return [[T1]] : !llvm.ptr<i32>
// CHECK-NEXT:  }
```
And the second case be:
```
// CHECK:  func.func @Memref2PointerIndex([[A0:%.*]]: memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i8> {
// CHECK-NEXT:    [[T0:%.*]] = "polygeist.memref2pointer"([[A0]]) : (memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<!llvm.struct<(i32, i32)>>
// CHECK-NEXT:    [[T1:%.*]] = llvm.getelementptr [[T0]][0, 1] : (!llvm.ptr<!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32>>
// CHECK-NEXT:    [[T2:%.*]] = llvm.bitcast([[T1]]) : !llvm.ptr<i32> to !llvm.ptr<i8>
// CHECK-NEXT:    return [[T2]] : !llvm.ptr<i8>
// CHECK-NEXT:  }
```
?

Reply:
Giving this a second though, I don't really see the point of having this canonicalization pattern in the first place. As you say, the lowering should be taking care of this. Do you think we should disable it altogether?

Reply:
The transformation that this PR address is `Memref2PointerIndex`:
```
/// Simplify pointer2memref(memref2pointer(x)) to cast(x)
class Memref2PointerIndex final : public OpRewritePattern<Memref2PointerOp> {
```
According to the comment this transformation should 'fold' cack to pointer2memref, memref2pointer operations into a conversion. The new test case does not have that pattern so I'm surprised it consider this test case a candidate at all. 
Is the transformation actually doing something else (the comment is misleading)?

Reply:
The comment is misleading, I forgot to point it out in a comment.

Reply:
Right, it should be more like
/// Simplify memref2pointer(subindex(x)) to getelementptr(memref2pointer(x))

Reply:
> Giving this a second though, I don't really see the point of having this canonicalization pattern in the first place. As you say, the lowering should be taking care of this. Do you think we should disable it altogether?

Right, I don't see the point too, removed. 😄

Reply:
hmm..ok, if everyone is ok, I will revert the last commit and do it in a separate PR.

Reply:
I'm OK with the fix.  I do think that this transformation should generate a `"polygeist.memref2pointer` with a result type that is a llvm.ptr<elemTy> where elemTy is identical to the original memref element type. If there is a need for a bitcast operation I think it should be injected after the GEP operation. That is not what this pass does, but in this PR I am happy we just fix the wrong codegen and leave the remaining consideration to the future (we might want to revisit all these polygeist canonicalization passes to make sure they make sense for our use cases).